### ISSUE TITLE
Better structure for the repo to accommodate different clients' needs

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-java-instrumentation#597]
Rename `agent-bootstrap` directory to `auto-bootstrap`, and rename `agent-tooling` directory to `auto-tooling`.

**Comment by iNikem** [comment]
Or rename module names? :) Why auto is better than agent?

**Comment by anuraaga** [comment]
I'm happy with any that make the folder / gradle task match up, including just `bootstrap` and `tooling`. The actual archive names generally aren't important during normal development so I've always found it annoying when the tasks don't match up with the folder names and I have to look up what the actual name is supposed to be in a file. Instead, using `archivesBaseName` to tweak the archives to be different from folder names has always been easier for me to use.

**Comment by anuraaga** [comment]
Let me add here instead of filing a new issue for now, it might be alleviated by naming or similar.

I currently find the distinction between bootstrap and tooling to be confusing. I'd expect classes like `TracerInstaller` to be in bootstrap, and others like `ClientDecorator` to be in tooling. I feel like every time I'm looking for a class I look in the wrong one first :)

Should more of tooling be moved into bootstrap and instrumentation stuff like `ClientDecorator` moved closer to `typed`, and we rename that from `tooling` to `instrumentation-common`?

**Comment by trask** [comment]
oh ya, `bootstrap` is stuff that goes in the bootstrap class loader, not stuff that is used for bootstrapping the agent

it would be nice for that to be more self-explanatory

some initial thoughts for a more descriptive name: `bootstrap-class-loader`, `bootstrap-cl`, `global`, `global-namespace`, ...

**Comment by trask** [comment]
trying to think about `auto-bootstrap` vs `agent-bootstrap`...

(setting aside temporarily whether/how to rename `bootstrap` and `tooling`...)

i think `auto` goes better with the current `groupId`, e.g.

* `io.opentelemetry.instrumentation.auto:opentelemetry-javaagent`
* `io.opentelemetry.instrumentation.auto:opentelemetry-auto-bootstrap`
* `io.opentelemetry.instrumentation.auto:opentelemetry-auto-tooling`

but that matches the pattern for instrumentation:

* `io.opentelemetry.instrumentation.auto:opentelemetry-auto-X`

which doesn't seem ideal.

what about having a third `groupId` for javaagent infrastructure `io.opentelemetry.javaagent`, e.g.

* `io.opentelemetry.javaagent:opentelemetry-javaagent`
* `io.opentelemetry.javaagent:opentelemetry-javaagent-bootstrap`
* `io.opentelemetry.javaagent:opentelemetry-javaagent-tooling`

one downside to this is it creates dependency cycle between the groupIds `io.opentelemetry.instrumentation.auto` and `io.opentelemetry.javaagent`, but i don't think that's too bad

**Comment by trask** [comment]
as far as naming for the `bootstrap` module, most of it looks like it would belong in an `instrumentation-api` module:
* io.opentelemetry.auto.bootstrap.instrumentation.**
* io.opentelemetry.auto.bootstrap.instrumentation.api.*
* io.opentelemetry.auto.bootstrap.config.Config
* io.opentelemetry.auto.bootstrap.CallDepthThreadLocalMap

and we need an `instrumentation-api` module anyways for people who want to bring their own instrumentation.

hm, but there's also part of "instrumentation api" in `tooling` (that lives in AgentClassLoader):
* io.opentelemetry.auto.tooling.bytebuddy.**

so maybe we need two "instrumentation api" modules:
* `instrumentation-runtime-api`
* `instrumentation-matcher-api`

then what's left in `bootstrap` could be `javaagent-bcl`

and what's left in `tooling` could be `javaagent-acl` (unfortunately sounds like security related "access control list")

these names are not clear on their own, but at least they are not misleading, like "bootstrap" which new contributors probably will interpret as "bootstrapping the agent" instead of "bootstrap class loader"

in summary (but definitely not sold on the proposal), something like:

* `io.opentelemetry.javaagent:opentelemetry-javaagent`
* `io.opentelemetry.javaagent:opentelemetry-javaagent-acl`
* `io.opentelemetry.javaagent:opentelemetry-javaagent-bcl`
* `io.opentelemetry.javaagent:opentelemetry-instrumentation-runtime-api`
* `io.opentelemetry.javaagent:opentelemetry-instrumentation-matcher-api`

**Comment by trask** [comment]
added to `v0.6.0` since we are discussing changes to published groupIds/artifactIds

**Comment by anuraaga** [comment]
Yup was also leaning towards splitting out more modules.

I think the instrumentation APIs can live with the instrumentation, so my idea would be

The agent
- io.opentelemetry.javaagent:opentelemetry-javaagent
- io.opentelemetry.javaagent:opentelemetry-javaagent-acl (don't know about these two names though unless we're going to add `ccl` >D)
- io.opentelemetry.javaagent:opentelemetry-javaagent-bcl

The components for instrumentation
- io.opentelemetry.instrumentation:opentelemetry-instrumentation-runtime-api
- io.opentelemetry.instrumentation.auto:opentelemetry-instrumentation-matcher-api

**Comment by iNikem** [comment]
Don't we overthink this too much? The distinction of `tooling` and `bootstrap` modules is described (quite well IMO :) ) in `CONTRIBUTING.md`. There are two separate modules for the sole reason of simplifying their packaging and separation into different classloaders. One of them is exposed to the actual host application, another is well hidden.

I don't see any need neither for a new groupId nor for more modules. We may want to rename existing two modules to better reflect their differences: one is visible to a host application, another is not. From this POV I think that `agent-tooling` is a good name (and better than `auto-tooling`). `bootstrap` indeed may be better off as `api`, `bridge`, `shim`, `shared` or some other.

**Comment by anuraaga** [comment]
I feel as if an app that uses manual instrumentation, possibly for security reasons, would be quite surprised to see an Agent packed into their app. There may be no real concern without the tooling package available for it to reflect into, but I can imagine users asking about this in the future.

Don't you think it's a clear separation of concerns between bootstrap classloader classes for loading the agent and the support classes for instrumentation, possibly manually loaded?

**Comment by trask** [comment]
I think having an "auto instrumentation api" artifact is helpful for people who want to write their own auto instrumentation.

It makes our public contract with them clearer (both for them and for us), and makes it less likely for them to depend on something that we consider internal.

**Comment by trask** [comment]
Also I think we should have a (separate) artifact for the stuff that manual instrumentation needs (e.g. base tracers).

This is our public contract with people who want to write their own manual instrumentation based on our tracers/etc, and also prevents manual instrumentation from having to pull in any auto-instrumentation related classes (agree this will likely be question from users if they see manual instrumentation depending on artifact with auto-instrumentation related classes).

**Comment by iNikem** [comment]
It seems that this task is no longer about “renaming” but full-blown restructuring of the codebase.

And at least we have a hard requirement: separate artifact for manual instrumentations which does not touch agent parts.

**Comment by trask** [comment]
Oh yes, good point, this issue has definitely meandered off course (on course?) 😂
